File: VotingApp_Client_WEB/HttpRequestHandler.js
// Define the RequestType enum
export const RequestType = {
    SignIn: 'SignIn',
    LogIn: 'LogIn',
    CreateSession: 'CreateSession',
    FetchSessions: 'FetchSessions',
    FetchSession: 'FetchSession',
    PostResults: 'PostResults',
    StartSession: 'StartSession',
    EndSession: 'EndSession',
    DeleteSession: 'DeleteSession',
    UpdateSession: 'UpdateSession'
};

const serverPort = "5010";

// Define the HttpRequestHandler class
export class HttpRequestHandler {
    // API Routes
    static createUserRoute = "http://localhost:"+serverPort+"/api/user/create"; //post
    static loginUserRoute = "http://localhost:"+serverPort+"/api/user/login"; //post
    static createSessionRoute = "http://localhost:"+serverPort+"/api/session"; //post
    static startSessionRoute = "http://localhost:"+serverPort+"/api/session/start/"; //post
    static endSessionRoute = "http://localhost:"+serverPort+"/api/session/end/"; //post
    static postResultsRoute = "http://localhost:"+serverPort+"/api/session/results/"; //post
    static fetchSessionsRoute = "http://localhost:"+serverPort+"/api/sessions/user"; //post
    static fetchSessionRoute = "http://localhost:"+serverPort+"/api/session/"; // get
    static deleteSessionRoute = "http://localhost:"+serverPort+"/api/session/"; //delete
    static updateSessionRoute = "http://localhost:"+serverPort+"/api/session/"; //put

    static async sendHttpRequestAsync(requestType, content, routeExtension) {
        // decide which route to use
        let route;
        switch(requestType) {
            case RequestType.SignIn:
                route = this.createUserRoute + routeExtension;
                break;
            case RequestType.LogIn:
                route = this.loginUserRoute + routeExtension;
                break;
            case RequestType.CreateSession:
                route = this.createSessionRoute + routeExtension;
                break;
            case RequestType.FetchSessions:
                route = this.fetchSessionsRoute + routeExtension;
                break;
            case RequestType.FetchSession:
                route = this.fetchSessionRoute + routeExtension;
                break;
            case RequestType.PostResults:
                route = this.postResultsRoute + routeExtension;
                break;
            case RequestType.StartSession:
                route = this.startSessionRoute + routeExtension;
                break;
            case RequestType.EndSession:
                route = this.endSessionRoute + routeExtension;
                break;
            case RequestType.DeleteSession:
                route = this.deleteSessionRoute + routeExtension;
                break;
            case RequestType.UpdateSession:
                route = this.updateSessionRoute + routeExtension;
                break;
            default:
                throw new Error("Unsupported request type: " + requestType);
        }

        // Configuring request options
        const requestOptions = {
            method: '',
            headers: {
                'Content-Type': 'application/json'
            },
            body: content
        };

        // Determine HTTP method based on request type
        switch (requestType) {
            case RequestType.SignIn:
                requestOptions.method = 'POST';
                break;
            case RequestType.LogIn:
                requestOptions.method = 'POST';
                break;
            case RequestType.CreateSession:
                requestOptions.method = 'POST';
                break;
            case RequestType.FetchSessions:
                requestOptions.method = 'POST';
                break;
            case RequestType.PostResults:
                requestOptions.method = 'POST';
                break;
            case RequestType.StartSession:
                requestOptions.method = 'POST';
                break;
            case RequestType.EndSession:
                requestOptions.method = 'POST';
                break;
            case RequestType.UpdateSession:
                requestOptions.method = 'PUT';
                break;
            case RequestType.FetchSession:
                requestOptions.method = 'GET';
                requestOptions.body = null;
                break;
            case RequestType.DeleteSession:
                requestOptions.method = 'DELETE';
                break;
            default:
                throw new Error('Unsupported request type: ' + requestType);
        }

        // Send the HTTP request
        try {
            console.log(content);
            console.log(requestOptions.body);
            const response = await fetch(route, requestOptions);
            return response;
        } catch (error) {
            throw new Error('Failed to fetch: ' + error.message);
        }
    }
}

File: VotingApp_Client_WPF/HttpRequestHandler.cs
using System;
using System.Net.Http;
using System.Text;
using System.Threading.Tasks;

namespace VotingApp_Client_WPF
{
    public enum RequestType
    {
        SignIn,
        LogIn,
        CreateSession,
        FetchSessions,
        FetchSession,
        PostResults,
        StartSession,
        EndSession,
        DeleteSession,
        UpdateSession
    }

    internal class HttpRequestHandler
    {
        private const string serverPort = "5010";
        // API - Routes
        //  Users
        private static string createUserRoute = "http://localhost:"+serverPort+"/api/user/create"; //post
        private static string loginUserRoute = "http://localhost:"+serverPort+"/api/user/login"; //post

        // Sessions
        private static string createSessionRoute = "http://localhost:"+serverPort+"/api/session"; //post
        private static string startSessionRoute = "http://localhost:"+serverPort+"/api/session/start/"; //post
        private static string endSessionRoute = "http://localhost:"+serverPort+"/api/session/end/"; //post
        private static string postResultsRoute = "http://localhost:"+serverPort+"/api/session/results/"; //post
        private static string fetchSessionsRoute = "http://localhost:"+serverPort+"/api/sessions/user"; //post
        private static string fetchSessionRoute = "http://localhost:"+serverPort+"/api/session/"; // get
        private static string deleteSessionRoute = "http://localhost:"+serverPort+"/api/session/"; //delete
        private static string updateSessionRoute = "http://localhost:"+serverPort+"/api/session/"; //put

        // HTTP - Client
        private static HttpClient _client = new();

        public static async Task<HttpResponseMessage> SendHttpRequestAsync(RequestType requestType, string content, string routeExtension)
        {
            // decide which route to use
            switch(requestType)
            {
                case RequestType.SignIn: return await SendPostRequestAsync(createUserRoute+routeExtension, content);
                case RequestType.LogIn: return await SendPostRequestAsync(loginUserRoute+routeExtension, content);
                case RequestType.CreateSession: return await SendPostRequestAsync(createSessionRoute+routeExtension, content);
                case RequestType.FetchSessions: return await SendPostRequestAsync(fetchSessionsRoute+routeExtension, content);
                case RequestType.FetchSession: return await SendGetRequestAsync(fetchSessionRoute+routeExtension);
                case RequestType.PostResults: return await SendPostRequestAsync(postResultsRoute+routeExtension, content);
                case RequestType.StartSession: return await SendPostRequestAsync(startSessionRoute+routeExtension, content);
                case RequestType.EndSession: return await SendPostRequestAsync(endSessionRoute+routeExtension, content);
                case RequestType.DeleteSession: return await SendDeleteRequestAsync(deleteSessionRoute+routeExtension, content);
                case RequestType.UpdateSession: return await SendPutRequestAsync(updateSessionRoute+routeExtension, content);
            }
            throw new NotSupportedException("Unsupported request type: " + requestType.ToString());
        }

        private static async Task<HttpResponseMessage> SendGetRequestAsync(string route)
        {
            // send get request
            return await _client.GetAsync(route);
        }

        private static async Task<HttpResponseMessage> SendPostRequestAsync(string route, string content)
        {
            // send post request
            return await _client.PostAsync(route, new StringContent(content, Encoding.UTF8, "application/json"));
        }

        private static async Task<HttpResponseMessage> SendDeleteRequestAsync(string route, string content)
        {
            // config delete request (_client.DeleteAsync() doesnt provide requestbody functionality)
            var request = new HttpRequestMessage
            {
                Method = HttpMethod.Delete,
                RequestUri = new Uri(route),
                Content = new StringContent(content, Encoding.UTF8, "application/json")
            };

            // send delete request
            return await _client.SendAsync(request);
        }

        private static async Task<HttpResponseMessage> SendPutRequestAsync(string route, string content)
        {
            // send delete request
            return await _client.PutAsync(route, new StringContent(content, Encoding.UTF8, "application/json"));
        }
    }
}
